### ISSUE TITLE
Twisted web client IPV6 DNS lookup failed

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8070]
|<img alt="stevedxor's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| stevedxor reported|
|-|-|
|Trac ID|trac#8070|
|Type|release blocker: regression|
|Created|2015-10-27 04:15:12Z|
|Branch|https://github.com/twisted/twisted/tree/web-client-ipv6-brackets-8070-2|

I'm getting a "DNS lookup failed: address '[::1]' not found: [Errno -2] Name or service not known" even after patched with https://github.com/twisted/twisted/commit/1c4f621548b5721ae69c4024b41bac3bb12d792c.

My code:

```
from twisted.trial import unittest   
from urllib import urlencode
from base64 import b64encode
from twisted.python.log import err
from twisted.web.client import Agent, readBody
from twisted.internet import reactor
from twisted.internet.ssl import ClientContextFactory
from twisted.web.http_headers import Headers
from zope.interface import implements
from twisted.internet.defer import succeed
from twisted.web.iweb import IBodyProducer


class StringProducer(object):
    implements(IBodyProducer)

    def __init__(self, body):
        self.body = body
        self.length = len(body)

    def startProducing(self, consumer):
        consumer.write(self.body)
        return succeed(None)

    def pauseProducing(self):
        pass

    def stopProducing(self):
        pass

class WebClientContextFactory(ClientContextFactory):
    def getContext(self, hostname, port):
        return ClientContextFactory.getContext(self)

class HttpsClientTestCases(unittest.TestCase):
    def test_https_client(self):
        def cbRequest(response):
            print 'Response version:', response.version
            print 'Response code:', response.code
            print 'Response phrase:', response.phrase
            print 'Response headers:[{}]'.format(list(response.headers.getAllRawHeaders()))
            d = readBody(response)
            d.addCallback(cbBody)
            return d

        def cbBody(body):
            print 'Response body:'
            print body

        contextFactory = WebClientContextFactory()
        agent = Agent(reactor, contextFactory)

        authorization = b64encode(b"username:password")
        data = StringProducer({'hello': 'world'})
        d = agent.request(
            'POST',
            'https://[::1]',
            headers = Headers(
                {
                    'Content-Type': ['application/x-www-form-urlencoded'],
                    'Authorization': ['Basic ' + authorization]
                }
            ),
            bodyProducer = data
        )

        d.addCallbacks(cbRequest, err)
        d.addCallback(lambda ignored: reactor.stop())
        return d
```

Callstack shows the error on the line "d = agent.request(".

Stackoverflow link:
http://stackoverflow.com/questions/33359368/twisted-web-client-ipv6-dns-lookup-failed

Thanks!
Error message is "DNS lookup failed: address '[::1]' not found: [Errno -2] Name or service not known"

I had tried different IPV6 addresses and got similar errors.

Attachments:

* [8070_1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/01e/01ef37944a78c3fe87a69f09ea611d5933c503d2/846c6c87ecf50d0d68f4a974413562b566acf29a.patch) (5070 bytes) - added by edorum2 on 2015-11-23 11:42:38Z - Patch for bracket problem
* [8070_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/01e/01ef37944a78c3fe87a69f09ea611d5933c503d2/1d9247fb7bf9d7270202343ad8abcba0d680fd42.patch) (5986 bytes) - added by edorum2 on 2015-11-24 10:35:56Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8070 8070
type__release_blocker__regression release blocker: regression
reporter__stevedxor stevedxor
priority__high high
milestone__Twisted_15_5 Twisted 15.5
branch__branches_web_client_ipv6_brackets_8070_2 branches/web-client-ipv6-brackets-8070-2
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1445919312422699 1445919312422699
changetime__1463650058660663 1463650058660663
version__None None
owner__edorum2 edorum2

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="posita's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|posita commented|
|-|-|

See also [#8034](https://github.com/twisted/twisted/issues/8034) and [#8035](https://github.com/twisted/twisted/issues/8035).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

[#8071](https://github.com/twisted/twisted/issues/8071) was a duplicate of this.

**Comment by twisted-trac** [comment]
|<img alt="edorum2's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|edorum2 commented|
|-|-|

My patch removes the brackets from the host field of an URI, when that is done the twisted.web.abstrach.isIPv6Address will return True on the host, and we don't get the problem described in this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

(In [46327]) Branching to web-client-ipv6-brackets-8070.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to edorum2|
|-|-|

Thanks for the patch.

Is this bug also present in Twisted 15.4.0 ?

The test and the docstring of a test should not be targeted to a specific bug, but rather describe the functionality/feature which is tested.

Instead of `Check that bug #8118 is fixed` we can have something like:

`IPv6 address literals with brackets are supported as defined in RFC 2732 section 2 and they can be serialized and deserialized`.

Maybe the news file should have a similar tone... instead of just "correctly" it can talk about RFC 2732 section 2 and then people should have a better understand of what is correct and what is not :)

Looking forward for the next patch :).

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban commented|
|-|-|

I have also sent your path to buildot, just to see if there are no side effects of which we are not aware :) ... but looks good so far.

**Comment by twisted-trac** [comment]
|<img alt="edorum2's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|edorum2 commented|
|-|-|

This bug was introduced with the fix for bug [#7650](https://github.com/twisted/twisted/issues/7650), and that was post 15.4.0.

I like your docstring and topfile better. Do you want a new patch with those documentation changes or do you do those changes on your end?

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban commented|
|-|-|

Please send a patch on top of the latest patch.

Also, please document the new brackethost in the URITests docstring.

If this code was not release then this is not a public bug. I think that the topfile should be just misc... otherwise, people will see a note in the release notes that the IPv6 address literal is now fixed in 15.5.0 and might think that it was broken in 15.4.0

Instead, you can move the content of the topfile as a comment insiede URI.__init__ so that other devs will know why the brackets are striped, without having to check all the tests... but I am not sure if this is a good idea :)

**Comment by twisted-trac** [comment]
|<img alt="edorum2's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|edorum2 commented|
|-|-|

Added a new patch with requested improvements.
In addition it changes bracketHost to uriHost, which I think is a better variable name.

**Comment by twisted-trac** [comment]
|<img alt="edorum2's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|edorum2 removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban commented|
|-|-|

(In [46332]) Branching to web-client-ipv6-brackets-8070-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban set owner to edorum2|
|-|-|

Changes looks good. Thanks.

The content of the .misc topfile will be ignored so there is no need to put any content in it.

I have extended the test to check that netloc will keep the brackets.

I have also updated a few docstrings.

There is not much help to include "check that" in the test docstring... as this is why we have all tests :) https://jml.io/pages/test-docstrings.html

Please check that my changes make sense.

If all is ok and tests pass (they should) I will merge.

Many thanks for reporting and fixing this bug!

**Comment by twisted-trac** [comment]
|<img alt="edorum2's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|edorum2 commented|
|-|-|

Your changes looks good to me.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46340]) Merge web-client-ipv6-brackets-8070-2: Fix IPv6 address literal representation in URL.

Author: edorum2
Reviewer: adiroiban
Fixes: [#8070](https://github.com/twisted/twisted/issues/8070)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_16"></a>|@hawkowl commented|
|-|-|

(In [46341]) Merge web-client-ipv6-brackets-8070-2: Fix IPv6 address literal representation in URL.

Author: edorum2
Reviewer: adiroiban
Fixes: [#8070](https://github.com/twisted/twisted/issues/8070)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_17"></a>|@hawkowl commented|
|-|-|

In changeset 2a6c7818ae18bd56ba2406d0a2f01b723bc5dc62
```
#!CommitTicketReference repository="" revision="2a6c7818ae18bd56ba2406d0a2f01b723bc5dc62"
Merge release-16.2.0-8270: Release Twisted 16.2

Author: hawkowl
Reviewer: markrwilliams
Fixes: #8070
```
